flexmeasures/data/models/generic_assets.py

```diff
@@ -498,6 +498,89 @@ def chart(
 
         return chart_specs
 
+    def multiple_chart(
+        self,
+        chart_type: str = "chart_for_multiple_sensors",
+        event_starts_after: datetime | None = None,
+        event_ends_before: datetime | None = None,
+        beliefs_after: datetime | None = None,
+        beliefs_before: datetime | None = None,
+        source: DataSource
+        | list[DataSource]
+        | int
+        | list[int]
+        | str
+        | list[str]
+        | None = None,
+        include_data: bool = False,
+        dataset_name: str | None = None,
+        resolution: str | timedelta | None = None,
+        **kwargs,
+    ) -> list[dict]:
+        """Create a vega-lite chart showing sensor data.
+
+        :param chart_type: currently only "bar_chart" # todo: where can we properly list the available chart types?
+        :param event_starts_after: only return beliefs about events that start after this datetime (inclusive)
+        :param event_ends_before: only return beliefs about events that end before this datetime (inclusive)
+        :param beliefs_after: only return beliefs formed after this datetime (inclusive)
+        :param beliefs_before: only return beliefs formed before this datetime (inclusive)
+        :param source: search only beliefs by this source (pass the DataSource, or its name or id) or list of sources
+        :param include_data: if True, include data in the chart, or if False, exclude data
+        :param dataset_name: optionally name the dataset used in the chart (the default name is sensor_<id>)
+        :param resolution: optionally set the resolution of data being displayed
+        :returns: JSON string defining vega-lite chart specs
+        """
+        sensors = flatten_unique(self.sensors_to_show)
+        for sensor in sensors:
+            sensor.sensor_type = sensor.get_attribute("sensor_type", sensor.name)
+
+        # Ensure all elements are lists as search_beliefs requires a list of sensors
+        nested_sensors = [
+            [e] if not isinstance(e, list) else e for e in self.sensors_to_show
+        ]
+
+        # Initialize an empty list to store chart_specs for each group of sensors
+        return_description = []
+
+        # Loop through each set of sensors to generate individual chart_specs
+        for sensors in nested_sensors:
```

Comment:
Can we put the content of this loop into a util function, as I see this as a 95% repetition from code in `chart()`. 

Then it would be much more readable to see what this function is doing.

Maybe name it `_gather_specs_and_data_for_chart()`. Place at the end of this module I guess, should be fine. Or there is a great place in the ui package for utilities like this, you decide.

Reply:
I'd advise to keep any utils within the `data` package to prevent circular imports. Normally we try to have the `ui` package import from the `data` package, but not vice versa.

Reply:
Good point

Reply:
okay there is a `utils.py` in `data` package should i put it there or i could keep it in `generic_assets.py` as a utilty function.

Reply:
I'd say keep it here in `generic_assets.py`.

Reply:
okay.

Reply:
I added _gather_specs_and_data_for_chart() as an internal utility method within generic_assets.py because it directly uses class attributes and other methods specific to this class. If it's preferred to have it outside the class, I can refactor it to accept additional parameters, though this might make the function a bit more complex.